facilitator/src/batch.rs

```diff
@@ -299,84 +300,97 @@ impl<'a, H: Header, P: Packet> BatchWriter<'a, H, P> {
             batch,
             transport,
             trace_id,
+            use_bogus_packet_file_digest: false,
         }
     }
 
     pub fn path(&self) -> String {
         self.transport.path()
     }
 
-    /// Encode the provided header into Avro, sign that representation with the
-    /// provided key and write the header into the batch. Returns the signature
-    /// on success.
-    pub fn put_header(&self, header: &H, key: &EcdsaKeyPair) -> Result<Signature> {
-        let mut transport_writer = self.transport.put(self.batch.header_key(), self.trace_id)?;
-        let mut header_buf = Vec::new();
-        header.write(&mut header_buf)?;
-
-        transport_writer.write_all(&header_buf)?;
-        transport_writer
-            .complete_upload()
-            .context("failed to complete batch header upload")?;
-
-        // TODO(brandon): per the advice of ring's doc comments, create a single SystemRandom &
-        // reuse it throughout the application rather than creating a new SystemRandom each time.
-        key.sign(&SystemRandom::new(), &header_buf)
-            .context("failed to sign header file")
+    /// Sets whether this BatchWriter will use a bogus value for the packet
+    /// file digest when constructing the header of a validation batch. This
+    /// is intended only for testing.
+    pub fn set_use_bogus_packet_file_digest(&mut self, bogus: bool) {
+        self.use_bogus_packet_file_digest = bogus;
     }
 
-    /// Creates an avro_rs::Writer and provides it to the caller-provided
-    /// function, which may then write arbitrarily many packets into it. The
-    /// Avro encoding of the packet will be digested while they are written.
-    /// The operation should return Ok(()) when it has finished successfully or
-    /// some Err() otherwise. packet_file_writer returns the digest of all the
-    /// content written by the operation.
-    pub fn packet_file_writer<F>(&self, operation: F) -> Result<Digest>
-    where
-        F: FnOnce(&mut Writer<&mut DigestWriter<&mut Box<dyn TransportWriter>>>) -> Result<()>,
-    {
+    pub fn write(
+        &self,
+        key: &BatchSigningKey,
+        mut header: H,
```

Comment:
Does this need to take the header by value, or can it be `&mut header: H`?

Reply:
It could be a mut ref, but I prefer to pass ownership since `write` needs to modify the header in what is (from the caller's POV) a somewhat arbitrary fashion, & it's not clear that there are "normal" (i.e. production) uses where we want to keep the header around afterwards.

For tests & other places that want/need to keep the header around, all headers can be cloned. WDYT?

Reply:
It could be a mut ref, but I have a slight preference for passing ownership: the header that's passed is effectively a "template" since `write` needs to modify it in a way that is strange/arbitrary to the caller (i.e. `write` needs to add a digest). By passing ownership, the details of how the header is modified aren't included in the contract for `write` since it's `write`'s header anyway.

[Also, no "normal" (i.e. production) uses of `write` need to keep the header around after calling `write`, so we don't lose anything by consuming the header. For tests that want to keep the header around to e.g. check that it is round-tripped appropriately, all headers can be cloned.]

Reply:
I'm OK with this if the intent is to consume the header.